ln-gateway/src/bin/ln_gateway.rs

```diff
@@ -115,21 +135,91 @@ async fn htlc_accepted_handler(
     plugin: Plugin<PluginState>,
     value: serde_json::Value,
 ) -> Result<serde_json::Value, Error> {
-    let (htlc_sender, htlc_receiver) = oneshot::channel::<Result<Preimage, LnGatewayError>>();
+    let (sender, receiver) = oneshot::channel::<Result<Preimage, LnGatewayError>>();
     let gw_sender = { plugin.state().lock().await.clone() };
     let htlc_accepted: HtlcAccepted = serde_json::from_value(value)?;
     gw_sender
-        .send(GatewayRequest::HtlcAccepted(htlc_accepted, htlc_sender))
+        .send(GatewayRequest::HtlcAccepted(htlc_accepted, sender))
         .await
         .expect("failed to send over channel");
-    let preimage = htlc_receiver.await.unwrap()?;
+    let preimage = receiver.await??;
 
     Ok(serde_json::json!({
       "result": "resolve",
       "payment_key": preimage,
     }))
 }
 
+async fn balance(
+    plugin: Plugin<PluginState>,
+    _v: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let (sender, receiver) = oneshot::channel::<Result<Amount, LnGatewayError>>();
+    let gw_sender = { plugin.state().lock().await.clone() };
+    gw_sender
+        .send(GatewayRequest::Balance(sender))
+        .await
+        .expect("failed to send over channel");
+    let amount = receiver.await.unwrap()?;
+    Ok(json!({ "balance_msat": amount.milli_sat }))
+}
+
+async fn address(
+    plugin: Plugin<PluginState>,
+    _value: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let (sender, receiver) = oneshot::channel::<Address>();
+    let gw_sender = { plugin.state().lock().await.clone() };
+    gw_sender
+        .send(GatewayRequest::DepositAddress(sender))
+        .await
+        .expect("failed to send over channel");
+    let address = receiver.await.expect("Failed to read from channel");
+    Ok(json!({ "address": address }))
+}
+
+#[derive(Deserialize)]
+struct Deposit(
+    TxOutProof,
+    #[serde(deserialize_with = "serde_hex_deserialize")] Transaction,
+);
+
+async fn deposit(
+    plugin: Plugin<PluginState>,
+    value: serde_json::Value,
+) -> Result<serde_json::Value, Error> {
+    let deposit: Deposit = serde_json::from_value(value)?;
+    let (sender, receiver) = oneshot::channel::<Result<TransactionId, LnGatewayError>>();
+    let gw_sender = { plugin.state().lock().await.clone() };
+    gw_sender
+        .send(GatewayRequest::Deposit(deposit.0, deposit.1, sender))
+        .await
+        .expect("failed to send over channel");
+    let txid = receiver.await.unwrap()?;
+    Ok(json!({ "fedimint_txid": txid.to_string() }))
```

Comment:
I wonder if I can build a nice higher order function to deduplicate that code :smile: